This PR contains updates to the ZooKeeper chart for Kubernetes 1.5

kow3ns opened the pull request:
Upgrades the chart to use StatefulSet instead of the deprecated PetSet
Integrates with the 1.5 version of the ZooKeeper contrib example
Externalizes ZooKeeper configuration to allow the chart to be tuned to
different use cases and environments
Adds a ConfigMap to store the configuration
Adds a PodDisruptionBudget to ensure that kubectl drain respects the
ensemble's quorum
Defaults the ZooKeeper log retention policy to be 1 hour instead of
forever
Configures the ensemble to log to stdout for integration with Kubernetes
log rolling
Adds pod anti-affinity to ensure the ensemble is deployed in a way that
it tolerates node failure

Commit:
Remove init containers from the chart
Upgrades the chart to use StatefulSet instead of the deprecated PetSet
Integrates with the 1.5 version of the ZooKeeper contrib example
Externalizes ZooKeeper configuration to allow the chart to be tuned to
different use cases and environments
Adds a ConfigMap to store the configuration
Adds a PodDisruptionBudget to ensure that kubectl drain respects the
ensemble's quorum
Defaults the ZooKeeper log retention policy to be 1 hour instead of
forever
Configures the ensemble to log to stdout for integration with Kubernetes
log rolling
Adds pod anti-affinity to ensure the ensemble is deployed in a way that
it tolerates node failure

kow3ns commented:
@k8s-bot e2e test this

Commit:
This time with templates!!!

lachie83 commented:
We need to get the e2e running on 1.5 so that this can be tested cc @viglesiasce

viglesiasce commented:
@k8s-bot e2e test this

viglesiasce commented:
Test cluster is now on 1.5.x

kow3ns commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/charts/359)
<!-- Reviewable:end -->

viglesiasce commented on incubator/zookeeper/Chart.yaml:
```diff
@@ -1,10 +1,12 @@
 name: zookeeper
 home: https://zookeeper.apache.org/
-version: 0.1.1
+version: 0.1.2
```
Let's make this 0.2.0 since its a breaking change.

viglesiasce replied:
Lets also move this to the stable folder. The only additional thing to do is to add a NOTES.txt file that tells users how they can connect to the ZK cluster from both inside and outside the cluster. The Jenkins chart has a good example.

viglesiasce commented on incubator/zookeeper/README.md:
```diff
@@ -1,100 +1,178 @@
+| `Cpu` | The amount of CPU to request. As ZooKeeper is not very CPU intensive, `2` is a good choice to start with. | `2` |
+| `Heap` | The amount of JVM heap that the ZooKeeper servers will use. As ZooKeeper stores all of its data in memory, this value should reflect the size of your working set. The JVM -Xms/-Xmx format is used. |`2G` |
+| `Memory` | The amount of memory to request. This value should be at least 2 GiB larger than `Heap` to avoid swapping. You many want to use `1.5 * Heap` for values larger than 2GiB. The Kubernetes format is used. |`2Gi` |
+| `Storage` | The amount of Storage to request. Even though ZooKeeper keeps is working set in memory, it logs all transactions, and periodically snapshots, to storage media. The amount of storage required will vary with your workload, working memory size, and log and snapshot retention policy. | `250Gi`|
```
250Gi is a bit aggressive for an out of the box value can we bring this into the 2 digit range.

viglesiasce reviewed (changes requested):
Awesome stuff @kow3ns!! Great detail in the README! A few changes requested.

enisoc commented:
Reviewed 4 of 4 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[incubator/zookeeper/Chart.yaml, line 3 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8rWF5N9kn8u4FGSJG:-Ka8rWF6LXd3aLXFUYB8:bsoedmr) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/Chart.yaml#L3)):*
> ```YAML
> name: zookeeper
> home: https://zookeeper.apache.org/
> version: 0.1.2
> ```

I think this should be at least a minor version bump. If it wasn't in incubator, I might even recommend considering it a fork because you're downgrading from ZK 3.5.x to 3.4.x and someone may still want to use 3.5.x alpha.

---

*[incubator/zookeeper/README.md, line 8 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8s2fk-XIBHdobH3eE:-Ka8s2fl7gNYqCvrMExS:b-mq0wme) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/README.md#L8)):*
> ```Markdown
>   
> ## Prerequisites
> * Kubernetes 1.5 with alpha APIs enable
> ```

Is anti-affinity the only alpha API used here? Since that's described as optional below, can the rest of this chart work without alpha APIs enabled? If so, it would be good to explain that here.

---

*[incubator/zookeeper/README.md, line 9 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8sMry8a8fs_J1uzfn:-Ka8sMry8a8fs_J1uzfo:bgbsrew) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/README.md#L9)):*
> ```Markdown
> ## Prerequisites
> * Kubernetes 1.5 with alpha APIs enable
> * PV support on the underlying infrastructure
> ```

And also a dynamic provisioner?

---

*[incubator/zookeeper/README.md, line 205 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8u1_Rlm6XuWeNPdIG:-Ka8u1_Sc_QO47tIESHK:b-9632hh) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/README.md#L205)):*
According to the [user guide](http://kubernetes.io/docs/user-guide/persistent-volumes/#writing-portable-configuration), there are two storage-class annotations that are both beta. One of them just happens to have the word "alpha" in its name:

> Despite the word alpha in the name, the code behind this annotation has beta level support.

---

*[incubator/zookeeper/templates/zookeeper.yaml, line 203 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8v5j_D6n9_Bz0RGP-:-Ka8v5j_D6n9_Bz0RGP0:b-vl32q9) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/templates/zookeeper.yaml#L203)):*
> ```YAML
>       name: datadir
>       annotations:
>         {{if .Values.StorageClass}}
> ```

On other charts, the reviewer recommended a pattern that lets the user configure either the alpha or the beta annotation:

https://github.com/kubernetes/charts/pull/361/commits/95626d44db84935804aa029a573c8c75be2090f3

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359)*
<!-- Sent from Reviewable.io -->

Commit:
Updates to address the enisoc and vigersiasce comments
Defalut storage is is now 50 Gi
Default cores is set to 1
All templates have been moved to separate files (when they where in a
single manifest delete errors occured in helm)
NOTES.txt has been added.
Storage annontation has been modified to use beta if the user has
specified a value or alpha default if they have not

kow3ns commented:
Review status: 5 of 14 files reviewed at latest revision, 7 unresolved discussions.

---

*[incubator/zookeeper/Chart.yaml, line 3 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8p6XXOpQbOVFTRdcH-r2-3:-Ka9zUso5lDr0OyoDBcT:b-896fix) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/Chart.yaml#L3)):*
<details><summary><i>Previously, viglesiasce (Vic Iglesias) wrote…</i></summary><blockquote>

Lets also move this to the stable folder. The only additional thing to do is to add a NOTES.txt file that tells users how they can connect to the ZK cluster from both inside and outside the cluster. The Jenkins chart has a good example.
</blockquote></details>

Done.

---

*[incubator/zookeeper/Chart.yaml, line 3 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8rWF5N9kn8u4FGSJG:-Ka9fRYZHJ7nNhX8XF7M:b-l0dop2) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/Chart.yaml#L3)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary><blockquote>

I think this should be at least a minor version bump. If it wasn't in incubator, I might even recommend considering it a fork because you're downgrading from ZK 3.5.x to 3.4.x and someone may still want to use 3.5.x alpha.
</blockquote></details>

I'm fine with bumping the version. However, to be clear, the previous version uses 3.4.x or 3.5.x. I'm aware of no software in the Apache ecosystem that actually uses 3.5.x. 

---

*[incubator/zookeeper/README.md, line 8 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8s2fk-XIBHdobH3eE:-Ka9gE8klo0yf1GW1uTW:b-896fix) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/README.md#L8)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary><blockquote>

Is anti-affinity the only alpha API used here? Since that's described as optional below, can the rest of this chart work without alpha APIs enabled? If so, it would be good to explain that here.
</blockquote></details>

Done.

---

*[incubator/zookeeper/README.md, line 9 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8sMry8a8fs_J1uzfn:-Ka9zX7AYSjpj4nX7ZBp:b-896fix) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/README.md#L9)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary><blockquote>

And also a dynamic provisioner?
</blockquote></details>

Done.

---

*[incubator/zookeeper/README.md, line 57 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8p6XXOpQbOVFTRdcI-r2-57:-Ka9d-UQW8aB4QUTKS6f:b-u6dy6) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/README.md#L57)):*
<details><summary><i>Previously, viglesiasce (Vic Iglesias) wrote…</i></summary><blockquote>

250Gi is a bit aggressive for an out of the box value can we bring this into the 2 digit range.
</blockquote></details>

Done. The only thing to mention is that small volume sizes can result in poor I/O performance depending on the cloud provider. I'm going to make a note of this in the comments.

---

*[incubator/zookeeper/README.md, line 205 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8u1_Rlm6XuWeNPdIG:-Ka9eycRo2Y3-xo4Djm6:b-896fix) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/README.md#L205)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary><blockquote>

According to the [user guide](http://kubernetes.io/docs/user-guide/persistent-volumes/#writing-portable-configuration), there are two storage-class annotations that are both beta. One of them just happens to have the word "alpha" in its name:

> Despite the word alpha in the name, the code behind this annotation has beta level support.
</blockquote></details>

Done.

---

*[incubator/zookeeper/templates/zookeeper.yaml, line 203 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8v5j_D6n9_Bz0RGP-:-Ka9eUaZf4WDiDxgL1hs:b7ijyfj) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/templates/zookeeper.yaml#L203)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary><blockquote>

On other charts, the reviewer recommended a pattern that lets the user configure either the alpha or the beta annotation:

https://github.com/kubernetes/charts/pull/361/commits/95626d44db84935804aa029a573c8c75be2090f3
</blockquote></details>

I think the correct way, as per the user guide for K8s, is to include the alpha annotation with default if the user provides nothing or to include the beta annotation if the user provides a value.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359)*
<!-- Sent from Reviewable.io -->

Commit:
Removing incubator and bumping version

kow3ns commented:
@k8s-bot e2e test this

Commit:
moving back to incubator as testing fails

Commit:
removing stable

kow3ns commented:
@viglesiasce @foxish @enisoc  Two issues.
.1 helm del --purge failed on the original chart. This is related to the fact that the manifest contained multiple objects, and separating the API objects resolves it. I will open an issue, but, if this is the intended behavior it will be a documentation issue or a feature request. If it is not then it is a bug.
.2 The build fails when I move the chart from incubator to stable. If everyone concurs, I'd like to complete this PR and merge to incubator, and then open a separate PR to migrate to stable.

kow3ns commented:
Reviewed 3 of 19 files at r3, 12 of 18 files at r4.
Review status: 7 of 9 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359)*
<!-- Sent from Reviewable.io -->

kow3ns commented:
Reviewed 1 of 7 files at r1, 4 of 18 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359)*
<!-- Sent from Reviewable.io -->

enisoc commented:
I've definitely used multiple objects per YAML file in Helm templates, and regular delete worked fine (I've never used `--purge` before). So I would call it a bug if you can't `delete --purge` such a release.

---

Reviewed 1 of 19 files at r3, 16 of 18 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[incubator/zookeeper/README.md, line 8 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8s2fk-XIBHdobH3eE:-KaE-BYYmzjc3aLm2Oqs:b577s5g) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/README.md#L8)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Did you make a change for this?

---

*[incubator/zookeeper/README.md, line 9 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8sMry8a8fs_J1uzfn:-KaE-CujHRLZY52aOlLc:b577s5g) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/README.md#L9)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Did you make a change for this?

---

*[incubator/zookeeper/README.md, line 205 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8u1_Rlm6XuWeNPdIG:-KaE-GMtORJrUnsMR3fi:b577s5g) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/README.md#L205)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Did you make a change for this?

---

*[incubator/zookeeper/templates/NOTES.txt, line 22 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-KaDz8S3ZWqKx8EtU3Mp:-KaDz8S3ZWqKx8EtU3Mq:bwsdrj6) ([raw file](https://github.com/kubernetes/charts/blob/eed9319d5a4dcedbb08cf60f73648c82b83998d7/incubator/zookeeper/templates/NOTES.txt#L22)):*
> ```
> failures. It is important that you do not use IP addresses and instead use the 
> DNS entries for the domain established by the StatefulSet's HeadlessService. 
> Your connection string should look like {{.Release.Name  | trunc 24 }}-0.{{ printf "hs-%s" .Release.Name | trunc 24 }}:{{.Values.ClientPort}},{{.Release.Name  | trunc 24 }}-1.{{ printf "hs-%s" .Release.Name | trunc 24 }}:{{.Values.ClientPort}},... .
> ```

I think you could use a template loop to generate the exact connection string based on the values.yaml (considering replica count).

---

*[incubator/zookeeper/templates/ss.yaml, line 146 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-KaE-Yd863IiOEWLCcGP:-KaE-Yd863IiOEWLCcGQ:b-bh0mpj) ([raw file](https://github.com/kubernetes/charts/blob/eed9319d5a4dcedbb08cf60f73648c82b83998d7/incubator/zookeeper/templates/ss.yaml#L146)):*
> ```YAML
>       annotations:
>         {{if .Values.StorageClass}}
>         volume.beta.kubernetes.io/storage-class: {{.Values.StorageClass}}
> ```

It's good practice to defensively add `| quote` to any user-supplied values.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-:-KaDxldInheTkxeGRNN-:b-q6ak2b)*
<!-- Sent from Reviewable.io -->

kow3ns commented:
*[incubator/zookeeper/README.md, line 9 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8sMry8a8fs_J1uzfn:-KaEHlW0rRr2Oi0WXmjr:b-896fix) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/README.md#L9)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary><blockquote>

Did you make a change for this?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359)*
<!-- Sent from Reviewable.io -->

Commit:
adding back modifications to preamble of README

kow3ns commented:
Reviewed 1 of 19 files at r3, 2 of 18 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[incubator/zookeeper/README.md, line 8 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8s2fk-XIBHdobH3eE:-KaEH_tv8C3ljvr3T68A:b-896fix) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/README.md#L8)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary><blockquote>

Did you make a change for this?
</blockquote></details>

Done.

---

*[incubator/zookeeper/README.md, line 9 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8sMry8a8fs_J1uzfn:-KaEJOEpHh5QU3UEtQEe:b-896fix) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/README.md#L9)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[incubator/zookeeper/README.md, line 205 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8u1_Rlm6XuWeNPdIG:-KaEJR77M26Xy4adXdW7:b-896fix) ([raw file](https://github.com/kubernetes/charts/blob/04b2253dc1b3b2684e971e00925e4364c442b219/incubator/zookeeper/README.md#L205)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary><blockquote>

Did you make a change for this?
</blockquote></details>

Done.

---

*[incubator/zookeeper/templates/NOTES.txt, line 22 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-KaDz8S3ZWqKx8EtU3Mp:-KaEI-w0ahkVCMuqZEgx:b-c2ecsq) ([raw file](https://github.com/kubernetes/charts/blob/eed9319d5a4dcedbb08cf60f73648c82b83998d7/incubator/zookeeper/templates/NOTES.txt#L22)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary><blockquote>

I think you could use a template loop to generate the exact connection string based on the values.yaml (considering replica count).
</blockquote></details>

The template looping structure produces a string that looks messy. Also the number of servers, and which servers to add to the string, are dependent on the users use case. Ultimately, this is just an example, and the specific configuration for the users client library will vary with the deployment and the library used.

---

*[incubator/zookeeper/templates/ss.yaml, line 146 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-KaE-Yd863IiOEWLCcGP:-KaEJRwi0aI0cFVyoz8g:b-896fix) ([raw file](https://github.com/kubernetes/charts/blob/eed9319d5a4dcedbb08cf60f73648c82b83998d7/incubator/zookeeper/templates/ss.yaml#L146)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary><blockquote>

It's good practice to defensively add `| quote` to any user-supplied values.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359)*
<!-- Sent from Reviewable.io -->

enisoc commented:
Reviewed 1 of 19 files at r3, 2 of 18 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[incubator/zookeeper/templates/NOTES.txt, line 22 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-KaDz8S3ZWqKx8EtU3Mp:-KaELl_7yGxE-xw_ghdh:bif6d2k) ([raw file](https://github.com/kubernetes/charts/blob/eed9319d5a4dcedbb08cf60f73648c82b83998d7/incubator/zookeeper/templates/NOTES.txt#L22)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

The template looping structure produces a string that looks messy. Also the number of servers, and which servers to add to the string, are dependent on the users use case. Ultimately, this is just an example, and the specific configuration for the users client library will vary with the deployment and the library used.
</blockquote></details>

I see. If it isn't always going to be the same format or set of servers, then I agree it isn't worth it to add a loop. For future reference though, you can make a template loop produce any pretty format you want if you use space-trimming delimiters.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359)*
<!-- Sent from Reviewable.io -->

enisoc commented:
LGTM

lachie83 commented on incubator/zookeeper/templates/pdb.yaml:
```diff
@@ -0,0 +1,16 @@
+  selector:
+    matchLabels:
+      component: "{{.Release.Name}}"
+  minAvailable: {{.Values.MinAvailable}}
```
This should be computed from `{{.Values.Servers}}` such the MinAvailable provides quorum across the cluster.

kow3ns commented:
*[incubator/zookeeper/templates/pdb.yaml, line 16 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-KaAvm9CLWrUPpeQ-ixR-r5-16:-KaKZfDfXO--B1sPwZTs:b-kcue8k) ([raw file](https://github.com/kubernetes/charts/blob/47e2ab31506d0e788911c67c123798fca0f67ddf/incubator/zookeeper/templates/pdb.yaml#L16)):*
<details><summary><i>Previously, lachie83 (Lachlan Evenson) wrote…</i></summary><blockquote>

This should be computed from `{{.Values.Servers}}` such the MinAvailable provides quorum across the cluster.
</blockquote></details>

If the goal is to choose a reasonable default for MinAvailable for  N servers, the users intention is more likely to be N-1 then (N/2) + 1, which constitutes an ensemble in an unweighted quorum, but, in any case, we can't really know the users intention.

 I think the correct design would be to make MinAvailable optional, default to minAvailable = Servers -1, and override this configuration if the user specifies an explicit value.

 My reasoning is as follows.
 
If a user specifies 5 Servers, the user's intention is usually to allow for one by one rolling node restarts, and to allow the ensemble to achieve quorum if an unexpected node failure occurs during the one by one rolling node restarts. Users do not generally deploy 5 servers to support two by two rolling restarts. 

If a user deploys 3 servers (N/2) + 1 = N-1, and we get the same behavior in any case.

If a user deploys 1 server,  N-1 = 0, and (N/2) + 1 = 1. This case is interesting because if you deploy ZooKeeper with 1 instance, high availability can not be your intention. This is likely to be a testing or development deployment. With N-1 you would not block drain operations, but you would loose availability (temporarily) during rolling restarts. With (N/2) + 1 your deployment would block drain operations, potentially preventing critical patches, but it would still not tolerate  unplanned node failures or restarts. I think N-1 is the right thing to do here (The user could still specify 1 if they explicitly wanted to block drain operations).

If a user deploys 7 or 11 servers (more than this and ZooKeepers performance begins to degrade), I don't think we can infer anything. It could be to support read fan out for a minimal write, heavy read workload, or it could be to improve availability based on unreliable infrastructure (in which case  N-1 is still a reasonable choice). Allowing the user to explicitly specify MinAvailable is a must have for deployments like this.

@enisoc @foxish @erictune @janetkuo Any thoughts on the above?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359)*
<!-- Sent from Reviewable.io -->

lachie83 commented:
Thank you very much for your detailed response @kow3ns. Given that we're using PDB along with Helm we have the opportunity to set a reasonable computed default that can be overriden (if the user want) but I think the default should ensure quorum across a cluster given the number of replicas.  I simply want to avoid users relying on this to be correct and using tools like `kubectl drain` and breaking the cluster.

foxish commented:
*[incubator/zookeeper/templates/pdb.yaml, line 16 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-KaAvm9CLWrUPpeQ-ixR-r5-16:-KaO7VW_nH_HZNzpy1sL:bpdvkv7) ([raw file](https://github.com/kubernetes/charts/blob/47e2ab31506d0e788911c67c123798fca0f67ddf/incubator/zookeeper/templates/pdb.yaml#L16)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

If the goal is to choose a reasonable default for MinAvailable for  N servers, the users intention is more likely to be N-1 then (N/2) + 1, which constitutes an ensemble in an unweighted quorum, but, in any case, we can't really know the users intention.

 I think the correct design would be to make MinAvailable optional, default to minAvailable = Servers -1, and override this configuration if the user specifies an explicit value.

 My reasoning is as follows.
 
If a user specifies 5 Servers, the user's intention is usually to allow for one by one rolling node restarts, and to allow the ensemble to achieve quorum if an unexpected node failure occurs during the one by one rolling node restarts. Users do not generally deploy 5 servers to support two by two rolling restarts. 

If a user deploys 3 servers (N/2) + 1 = N-1, and we get the same behavior in any case.

If a user deploys 1 server,  N-1 = 0, and (N/2) + 1 = 1. This case is interesting because if you deploy ZooKeeper with 1 instance, high availability can not be your intention. This is likely to be a testing or development deployment. With N-1 you would not block drain operations, but you would loose availability (temporarily) during rolling restarts. With (N/2) + 1 your deployment would block drain operations, potentially preventing critical patches, but it would still not tolerate  unplanned node failures or restarts. I think N-1 is the right thing to do here (The user could still specify 1 if they explicitly wanted to block drain operations).

If a user deploys 7 or 11 servers (more than this and ZooKeepers performance begins to degrade), I don't think we can infer anything. It could be to support read fan out for a minimal write, heavy read workload, or it could be improve availability based on unreliable infrastructure (in which case  N-1 is still a reasonable choice). Allowing the user to explicitly specify MinAvailable is a must have for deployments like this.

</blockquote></details>

N-1 sounds reasonable to me, with the reasoning about 1 instance clusters. Since scaling isn't something we do with ZK, it should be fine. For other applications, we should support helm upgrade completely with StatefulSets which would update the PDB along with it when we scale.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359)*
<!-- Sent from Reviewable.io -->

kow3ns commented:
So are you okay with N-1 as the reasonable default or are you suggesting something else?

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-:-KaOBfKCi2KKRO5Wh81j:b-6qvdtw)*
<!-- Sent from Reviewable.io -->

lachie83 commented:
@kow3ns absolutely!

Commit:
Defaults MinAvailable to .Values.Servers-1. Cleans up some space
formatting in the template files.

kow3ns commented:
Review status: 5 of 9 files reviewed at latest revision, 4 unresolved discussions.

---

*[incubator/zookeeper/templates/pdb.yaml, line 16 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-KaAvm9CLWrUPpeQ-ixR-r5-16:-KbBa8LV_vICtziwIszd:b-896fix) ([raw file](https://github.com/kubernetes/charts/blob/47e2ab31506d0e788911c67c123798fca0f67ddf/incubator/zookeeper/templates/pdb.yaml#L16)):*
<details><summary><i>Previously, foxish (Anirudh Ramanathan) wrote…</i></summary><blockquote>

N-1 sounds reasonable to me, with the reasoning about 1 instance clusters. Since scaling isn't something we do with ZK, it should be fine. For other applications, we should support helm upgrade completely with StatefulSets which would update the PDB along with it when we scale.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359)*
<!-- Sent from Reviewable.io -->

viglesiasce commented:
@kow3ns anything left to do here? Is the n-1 change in yet? Do we want to move this chart to stable now too?

kow3ns commented:
LGTM, the N-1 is in and templated as discussed below. I was thinking of getting it into incubator under this PR and then moving it to stable under another?

---

Reviewed 1 of 19 files at r3, 4 of 18 files at r4, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-:-KbGoq1RCyNOJcJByFfC:b209mem)*
<!-- Sent from Reviewable.io -->

kow3ns commented:
@viglesiasce ^^

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-:-KbHBjeUs7COQWu6k8nT:b7ht4v8)*
<!-- Sent from Reviewable.io -->

Commit:
Merge branch 'master' of github.com:kubernetes/charts into zk-updates

viglesiasce commented on incubator/zookeeper/templates/ss.yaml:
```diff
@@ -0,0 +1,154 @@
+apiVersion: apps/v1beta1
+kind: StatefulSet
+metadata:
+  name: "{{.Release.Name  | trunc 24 }}"
```
Do you mind if I add the suffix zk to this? Otherwise its hard to figure out which pods belong to ZK within the release.

viglesiasce commented on incubator/zookeeper/values.yaml:
```diff
@@ -2,16 +2,23 @@
+Memory: "4Gi"
+Heap: "2G"
+Storage: "50Gi"
+StorageClass : "default"
```
Needs to be uncommented. Otherwise PVCs arent being bound by default (at least in GKE).

viglesiasce reviewed (changes requested):
Tested with the changes I proposed and it worked as expected. I can make the updates if youd like @kow3ns.

kow3ns commented:
*[incubator/zookeeper/templates/ss.yaml, line 4 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-KaAvm9CLWrUPpeQ-ixS-r7-4:-Kbxaac4J92wfMmyoWc_:bjqmig6) ([raw file](https://github.com/kubernetes/charts/blob/5144d2b43bf776ba42f10bd1b2dd02fe5870e303/incubator/zookeeper/templates/ss.yaml#L4)):*
<details><summary><i>Previously, viglesiasce (Vic Iglesias) wrote…</i></summary><blockquote>

Do you mind if I add the suffix zk to this? Otherwise its hard to figure out which pods belong to ZK within the release. 
</blockquote></details>

not at all

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359)*
<!-- Sent from Reviewable.io -->

Commit:
Adds zk prefix to statefulset name
Removes quotes from default storage type

kow3ns commented:
Reviewed 1 of 19 files at r3, 2 of 18 files at r4, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[incubator/zookeeper/values.yaml, line 10 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8p6XXOpQbOVFTRdcK-r7-10:-Kc-1EdOsaLgsECVt_uE:b-896fix) ([raw file](https://github.com/kubernetes/charts/blob/5144d2b43bf776ba42f10bd1b2dd02fe5870e303/incubator/zookeeper/values.yaml#L10)):*
<details><summary><i>Previously, viglesiasce (Vic Iglesias) wrote…</i></summary><blockquote>

Needs to be uncommented. Otherwise PVCs arent being bound by default (at least in GKE).
</blockquote></details>

Done.

---

*[incubator/zookeeper/templates/ss.yaml, line 4 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-KaAvm9CLWrUPpeQ-ixS-r7-4:-Kc-1IzgXGk3WDL97IQs:b-896fix) ([raw file](https://github.com/kubernetes/charts/blob/5144d2b43bf776ba42f10bd1b2dd02fe5870e303/incubator/zookeeper/templates/ss.yaml#L4)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

not at all
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359)*
<!-- Sent from Reviewable.io -->

viglesiasce commented:
@kow3ns i just noticed that the other resources need the suffix too (pdb, configmap, etc) sorry for the churn.

viglesiasce commented:
*[incubator/zookeeper/values.yaml, line 10 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8p6XXOpQbOVFTRdcK-r7-10:-Kc-3-DUXG80WUwElKDk:bu49vmb) ([raw file](https://github.com/kubernetes/charts/blob/5144d2b43bf776ba42f10bd1b2dd02fe5870e303/incubator/zookeeper/values.yaml#L10)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I am terrible. I said uncommented but meant commented. I apologize for my idiocy.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359)*
<!-- Sent from Reviewable.io -->

Commit:
add prefixes to make Objects easier to find
comment out default storage class

kow3ns commented:
Done

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[incubator/zookeeper/values.yaml, line 10 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-Ka8p6XXOpQbOVFTRdcK-r7-10:-Kc-7QL5UaiPxGAp_NNy:b-896fix) ([raw file](https://github.com/kubernetes/charts/blob/5144d2b43bf776ba42f10bd1b2dd02fe5870e303/incubator/zookeeper/values.yaml#L10)):*
<details><summary><i>Previously, viglesiasce (Vic Iglesias) wrote…</i></summary><blockquote>

I am terrible. I said uncommented but meant commented. I apologize for my idiocy.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-:-Kc-7RgFu_jXOo7NMjPU:bb6un1j)*
<!-- Sent from Reviewable.io -->

kow3ns commented:
Reviewed 1 of 19 files at r3, 5 of 18 files at r4, 5 of 5 files at r9.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359)*
<!-- Sent from Reviewable.io -->

tazjin commented:
I'm testing this updated StatefulSet in a test cluster and am running into an issue that stops me from upgrading this release, for example with new values:

`Error: UPGRADE FAILED: PodDisruptionBudget.policy "pd-zk-lab-cluster" is invalid: spec: Forbidden: updates to poddisruptionbudget spec are forbidden.`

Not sure if there's some easy workaround that can be applied here, or why this field is updated at all (cluster size hasn't changed).

kow3ns commented:
What are you trying to do specifically?

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-:-Kc3sb-9QJBi4mLaGrW5:bcc144w)*
<!-- Sent from Reviewable.io -->

kow3ns commented:
My guess would be that 
```yaml
  {{- if .Values.MinAvailable}}
  minAvailable: {{.Values.MinAvailable}}
  {{- else}}
  minAvailable: {{sub .Values.Servers 1}}
  {{- end}}
```
is causing Helm to attempt to update the manifest even though the number of servers has not changed. And the PDB is requiring you to delete it first and then recreate it rather than attempt to patch the Spec, but I need the specific sequence of commands that you are running to verify.

---







*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359#-:-Kc3tXJFdBMOTJiTEg-h:b-f6oc47)*
<!-- Sent from Reviewable.io -->

viglesiasce commented:
Thanks @kow3ns!!

kow3ns commented:
Reviewed 1 of 18 files at r4, 1 of 4 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/charts/359)*
<!-- Sent from Reviewable.io -->

tazjin commented:
@kow3ns Happens to me when upgrading any chart that depends on this one. Probably not something you can fix here so I'll go and file this issue to the Helm project.

devth commented:
Hey @lachie83, I've observed that ZK does not update DNS resolution when pods get recycled and their IP addresses change even though this uses DNS. Does the liveness probe handle this, and have you considered this issue?
